src/fsharp/PrettyNaming.fs

```diff
@@ -109,66 +109,160 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
-        ( ']', "RBrack"); ]
+        ( ']', "RBrack"); |]
 
-    let private opCharDict = 
-        let t = new Dictionary<_,_>()
-        for (c,_) in opCharTranslateTable do 
-            t.Add(c,1)
+    /// The set of characters usable in custom operators.
+    let private opCharSet =
+        let t = new HashSet<_>()
+        for (c,_) in opCharTranslateTable do
+            t.Add(c) |> ignore
         t
         
-    let IsOpName (n:string) =
-        let rec loop i = (i < n.Length && (opCharDict.ContainsKey(n.[i]) || loop (i+1)))
+    let IsOpName (name:string) =
+        let nameLen = name.Length
+        let rec loop i = (i < nameLen && (opCharSet.Contains(name.[i]) || loop (i+1)))
         loop 0
 
+    let IsMangledOpName (n:string) =
+        n.StartsWith (opNamePrefix, System.StringComparison.Ordinal)
+
+    // +++ GLOBAL STATE
+    /// Compiles a custom operator into a mangled operator name.
+    /// For example, "!%" becomes "op_DereferencePercent"
+    let private compileCustomOpName =
+        let t2 =
+            let t2 = Dictionary<_,_> (opCharTranslateTable.Length)
+            for x, y in opCharTranslateTable do
+                t2.Add (x, y)
+            t2
+        /// The maximum length of the name for a custom operator character.
+        /// This value is used when initializing StringBuilders to avoid resizing.
+        let maxOperatorNameLength =
+            opCharTranslateTable
+            |> Array.maxBy (snd >> String.length)
+            |> snd
+            |> String.length
+
+        /// Memoize compilation of custom operators.
+        /// They're typically used more than once so this avoids some CPU and GC overhead.
+        let compiledOperators = Dictionary<_,_> (System.StringComparer.Ordinal)
+
+        fun op ->
+            // Has this operator already been compiled?
+            let found =
+                lock compiledOperators <| fun () ->
+                    match compiledOperators.TryGetValue op with
+                    | true, res -> Some res
+                    | false, _ -> None
+
+            match found with
+            | Some res -> res
+            | None ->
+                let sb = new System.Text.StringBuilder (opNamePrefix, opNamePrefix.Length + (op.Length * maxOperatorNameLength))
+                for c in op do
+                    match t2.TryGetValue c with
+                    | true, x ->
+                        sb.Append(x) |> ignore
+                    | false, _ ->
+                        sb.Append(c) |> ignore
+
+                /// The compiled (mangled) operator name.
+                let opName = sb.ToString ()
+
+                // Cache the compiled name so it can be reused.
+                lock compiledOperators <| fun () ->
+                    compiledOperators.[op] <- opName
+
+                opName
+
+    // +++ GLOBAL STATE
+    /// Compiles a custom operator into a mangled operator name.
+    /// For example, "!%" becomes "op_DereferencePercent"
```

Comment:
This is the exact same comment as on `compileCustomOpName` - should make it clear how the behaviors/assumptions differ